### PR TITLE
[FLINK-19446][canal][json] canal-json has a situation that -U and +U are equal, when updating the null field to be non-null

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*When `CanalJsonDeserializationSchema` deserializes the null field to be non-null, canal-json has a situation that -U and +U are equal. Because `deserializes` method has no distinction between null and empty field. `CanalJsonDeserializationSchema` distinguishes between null and empty field by field names through parsing `JsonNode`.*

## Brief change log

  - *`CanalJsonDeserializationSchema` deserializes the record for distinguishing between null and empty field by field names with parsing `JsonNode`.*

## Verifying this change

  - *Modify `canal-data-filter-table.txt` to build a test case which has a situation that canal-json has null and empty field in `old` field to verify the distinction between null and empty field.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by SteNicholas** [comment]
@wuchong , could you please help to review this simple pull request if you are available?

File: flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/canal/CanalJsonDeserializationSchema.java [hunk 1]
```diff
@@ -238,8 +244,9 @@ public void deserialize(@Nullable byte[] message, Collector<RowData> out) throws
                     // the underlying JSON deserialization schema always produce GenericRowData.
                     GenericRowData after = (GenericRowData) data.getRow(i, fieldCount);
                     GenericRowData before = (GenericRowData) old.getRow(i, fieldCount);
+                    final JsonNode oldField = root.get(FIELD_OLD);
                     for (int f = 0; f < fieldCount; f++) {
-                        if (before.isNullAt(f)) {
+                        if (before.isNullAt(f) && oldField.findValue(fieldNames.get(f)) == null) {
                             // not null fields in "old" (before) means the fields are changed
                             // null/empty fields in "old" (before) means the fields are not changed
```
Comment (wuchong) [inline flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/canal/CanalJsonDeserializationSchema.java:251]:
```suggestion
                            // fields in "old" (before) means the fields are changed
                            // fields not in "old" (before) means the fields are not changed
```

File: flink-formats/flink-json/src/test/resources/canal-data-filter-table.txt [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {"data":[{"id":"101","name":"scooter","description":"Small 2-wheel scooter","weight":"3.14"},{"id":"102","name":"car battery","description":"12V car battery","weight":"8.1"},{"id":"103","name":"12-pack drill bits","description":"12-pack of drill bits with sizes ranging from #40 to #3","weight":"0.8"},{"id":"104","name":"hammer","description":"12oz carpenter's hammer","weight":"0.75"},{"id":"105","name":"hammer","description":"14oz carpenter's hammer","weight":"0.875"},{"id":"106","name":"hammer","description":"16oz carpenter's hammer","weight":"1.0"},{"id":"107","name":"rocks","description":"box of assorted rocks","weight":"5.3"},{"id":"108","name":"jacket","description":"water resistent black wind breaker","weight":"0.1"},{"id":"109","name":"spare tire","description":"24 inch spare tire","weight":"22.2"}],"database":"mydb","es":1598944132000,"id":1,"isDdl":false,"mysqlType":{"id":"int(11)","name":"varchar(255)","description":"varchar(512)","weight":"float"},"old":null,"pkNames":["id"],"sql":"","sqlType":{"id":4,"name":12,"description":12,"weight":7},"table":"product","ts":1598944146308,"type":"INSERT"}
```
Comment (wuchong) [inline flink-formats/flink-json/src/test/resources/canal-data-filter-table.txt:1]:
```suggestion
{"data":[{"id":"101","name":"scooter","description":"Small 2-wheel scooter","weight":"3.14"},{"id":"102","name":"car battery","description":"12V car battery","weight":"8.1"},{"id":"103","name":"12-pack drill bits","description":"12-pack of drill bits with sizes ranging from #40 to #3","weight":"0.8"},{"id":"104","name":"hammer","description":"12oz carpenter's hammer","weight":"0.75"},{"id":"105","name":"hammer","description":"14oz carpenter's hammer","weight":"0.875"},{"id":"106","name":"hammer","description":null,"weight":"1.0"},{"id":"107","name":"rocks","description":"box of assorted rocks","weight":"5.3"},{"id":"108","name":"jacket","description":"water resistent black wind breaker","weight":"0.1"},{"id":"109","name":"spare tire","description":"24 inch spare tire","weight":"22.2"}],"database":"mydb","es":1598944132000,"id":1,"isDdl":false,"mysqlType":{"id":"int(11)","name":"varchar(255)","description":"varchar(512)","weight":"float"},"old":null,"pkNames":["id"],"sql":"","sqlType":{"id":4,"name":12,"description":12,"weight":7},"table":"product","ts":1598944146308,"type":"INSERT"}
```

I think you want to test the original `description` is null.

File: flink-formats/flink-json/src/test/resources/canal-data.txt [hunk 3]
```diff
@@ -1,5 +1,5 @@
 {"data":[{"id":"101","name":"scooter","description":"Small 2-wheel scooter","weight":"3.14"},{"id":"102","name":"car battery","description":"12V car battery","weight":"8.1"},{"id":"103","name":"12-pack drill bits","description":"12-pack of drill bits with sizes ranging from #40 to #3","weight":"0.8"},{"id":"104","name":"hammer","description":"12oz carpenter's hammer","weight":"0.75"},{"id":"105","name":"hammer","description":"14oz carpenter's hammer","weight":"0.875"},{"id":"106","name":"hammer","description":"16oz carpenter's hammer","weight":"1.0"},{"id":"107","name":"rocks","description":"box of assorted rocks","weight":"5.3"},{"id":"108","name":"jacket","description":"water resistent black wind breaker","weight":"0.1"},{"id":"109","name":"spare tire","description":"24 inch spare tire","weight":"22.2"}],"database":"inventory","es":1589373515000,"id":3,"isDdl":false,"mysqlType":{"id":"INTEGER","name":"VARCHAR(255)","description":"VARCHAR(512)","weight":"FLOAT"},"old":null,"pkNames":["id"],"sql":"","sqlType":{"id":4,"name":12,"description":12,"weight":7},"table":"products2","ts":1589373515477,"type":"INSERT"}
```
Comment (wuchong) [inline flink-formats/flink-json/src/test/resources/canal-data.txt:1]: ditto.

**Review by wuchong — CHANGES_REQUESTED** [review]
Thanks for the contribution. I left some comments.

**Comment by SteNicholas** [comment]
@wuchong , thanks for your detailed review. I have followed your comments. Please help to review again.

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**